Домашнее задание спринт №9

KravtsovSO64 opened the pull request:
Как смог восстановить ветку main до сданной ДЗ спринт №8. Сейчас глянул на пул dev in main. Вроде бы все верные изменения, ничего лишнего!

Commit:
Постарался исправить все замечания!

Commit:
Новые изменения

Commit:
Домашнее задание №9

faritowich commented on app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt:
```diff
@@ -1,14 +1,53 @@
+        return if (s.isNullOrEmpty()) {
+            View.GONE
+        } else {
+            View.VISIBLE
```
Для изменения видимости элемента на экране в данном случае удобнее использовать extension View.isVisible:
View.isVisible = true соответствует View.visibility = View.VISIBLE
View.isVisible = false соответствует View.visibility = View.GONE

faritowich commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,61 @@
+    </style>
+
+    <style name="Search_Line">
+        <item name="android:layout_width">328dp</item>
```
Конкретная ширина на нужна. У разных телефонов разные ширины экранов. match_parent нужен

KravtsovSO64 replied:
Убрал из Style Search_Line размеры и отступы, оставил атрибуты визуалки. Ширину принял за 0dp и задал EdityText вес, для корректного отображения кнопки сброса введенного текста. Теперь на любом экране корректно отображается кнопка сброса текста, если поле ввода не пусто.

faritowich commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,61 @@
+    <style name="SettingViewButtonText" >
+        <item name="android:layout_width">match_parent</item>
+        <item name="android:layout_height">wrap_content</item>
+        <item name="android:layout_gravity">start</item>
```
Лучше внешние отступы, высоту и ширину, веса, гравити не выносить в стиль. Если на экране будет много элементов, будет сложнее ориентироваться с их взаимным расположением, когда отвечающие за это атрибуты разнесены по разным файлам - что-то может быть в лейауте, что-то в стиле. В стилях обычно содержат атрибуты, отвечающие за внешний вид вьюхи и вообще только за то, что находится внутри нее - цвета, шрифты, фоны и т.д. Бывают исключения, когда, например, мы точно знаем, что у кнопки всегда должна строго определенная высота и ширина - тогда да, имеет смысл выносить

KravtsovSO64 replied:
Исправил создание общих стилей для атрибутов на Main  и Setting экранах.  
Стили MainViewButton и SettingViewButton теперь содержат атрибуты отвечающие за внешний вид. Размеры и отступы вынес в разметку кнопок. В кнопки на экране Setting перенес размеры и отступы иконок, единственный общий тег это цвет, решил, что не стоит из-за этого целый стиль создавать.

Commit:
v9.1 (Исправла поисковая строка, стили, сохранение данных)

KravtsovSO64 commented:
Исправил разметку EditeText в поисковой строке, теперь кнопка сброса введенного текста отображается и работает корректно на разных экранах и ориентации экрана.
Добавил сохраннее данных поискового запроса при изменение ориентации экрана. В манифесте на экрана добавил запрет на изменение ориентации, оставил только у экрана "Поиск" для проверки сохранения данных при повороте экрана. 
В стилях исправил замечания. Добавил наследование для кнопок на главном экране.
Протестировал приложение на эмуляторах и на телефоне работает корректно.

Commit:
v9.2 (Исправлы поисковая строка, стили, сохранение данных) + добавил lineSearchLine.setText(searchRequest) в методе onRestoreInstanceState

KravtsovSO64 commented:
Упустил добавление сохраненого поискового запроса в EditText. 
Добавил lineSearchLine.setText(searchRequest) в методе onRestoreInstanceState.

faritowich commented on app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt:
```diff
@@ -1,14 +1,74 @@
+
+    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
+        super.onRestoreInstanceState(savedInstanceState)
+        val lineSearchLine = findViewById<EditText>(R.id.editText)
```
Сейчас для вью R.id.editText переменная объявляется дважды в разных методах. Этого можно избежать, если объявлять глобальную переменную в теле класса с lateinit (будет private lateinit var viewName: EditText) или как нуллабл переменную (будет private var viewName: EditText? = null, а в onCreate уже задавать значение findViewById<EditText>(R.id.VIEW_ID). Все остальные переменные, отвечающие за вьюхи, можно объявить так же. Тогда переменную можно инициализировать один раз в onCreate и иметь к ней доступ из любой функции класса.

faritowich replied:
А вообще рекомендую почитать про библиотеку ViewBinding от гугла и использовать её - она используется повсеместно

faritowich commented on app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt:
```diff
@@ -1,14 +1,74 @@
+    companion object{
+        const val SEARCH_REQUEST = "SEARCH_REQUEST"
+        const val AMOUNT_DEF = ""
+    }
```
companion object принято объявлять в конце класса: https://kotlinlang.org/docs/coding-conventions.html#class-layout

faritowich commented on app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt:
```diff
@@ -1,14 +1,74 @@
+
+    companion object{
+        const val SEARCH_REQUEST = "SEARCH_REQUEST"
+        const val AMOUNT_DEF = ""
```
private

faritowich commented on app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt:
```diff
@@ -1,14 +1,74 @@
+        }
+    }
+
+    private var searchRequest : String = ""
```
лучше вынести в начало класса

Commit:
v9.3 исправлено:
- при удалении поискового запроса, убирается клавиатура;
- убраны тени у кнопок;
- в SearchActivity все переменные тепрь глобальные, для исклучения повторного объявления

Commit:
v9.5 небольшой рефакторинг кода
